eventstore/store_consolidated_billable_events.go

```diff
@@ -0,0 +1,319 @@
+package eventstore
+
+import (
+	"database/sql"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"code.cloudfoundry.org/lager"
+	"github.com/alphagov/paas-billing/eventio"
+)
+
+const (
+	DefaultConsolidationStartDate = "2017-07-01"
+)
+
+func (e *EventStore) GetConsolidatedBillableEventRows(filter eventio.EventFilter) (eventio.BillableEventRows, error) {
+	tx, err := e.db.Begin()
```

Comment:
mmm you are right. 

It is also wrong here: https://github.com/alphagov/paas-billing/blob/66eb9d4add24d3054c8f8712c5f9d543bebf5db2/eventstore/store_billable_events.go#L18-L30

Reply:
Adding the `tx.Commit()` blows it up

```diff
diff --git a/eventstore/store_billable_events.go b/eventstore/store_billable_events.go
index 75e35b6..6a490c1 100644
--- a/eventstore/store_billable_events.go
+++ b/eventstore/store_billable_events.go
@@ -25,7 +25,7 @@ func (s *EventStore) GetBillableEventRows(filter eventio.EventFilter) (eventio.B
                tx.Rollback()
                return nil, err
        }
-       return rows, nil
+       return rows, tx.Commit()
 }
 
 func (s *EventStore) getBillableEventRows(tx *sql.Tx, filter eventio.EventFilter) (eventio.BillableEven
diff --git a/eventstore/store_consolidated_billable_events.go b/eventstore/store_consolidated_billable_e
index c1480a5..06f65c8 100644
--- a/eventstore/store_consolidated_billable_events.go
+++ b/eventstore/store_consolidated_billable_events.go
@@ -25,7 +25,7 @@ func (e *EventStore) GetConsolidatedBillableEventRows(filter eventio.EventFilter
                tx.Rollback()
                return nil, err
        }
-       return rows, nil
+       return rows, tx.Commit()
 }
````

We get a lot of errors like:

```
      <*errors.errorString | 0xc42032f640>: {
          s: "unexpected command tag SELECT",
      } 
```

Those two methods are readonly, so I guess it makes sense to not create a transaction? :thinking:

Reply:
OK, now I know what is going on.

The object `BillableEventsRows` has a `Close()` that does the `tx.Rollback()` to avoid leaking transactions. The new and the old method use the same logic:

https://github.com/alphagov/paas-billing/blob/66eb9d4add24d3054c8f8712c5f9d543bebf5db2/eventstore/store_billable_events.go#L93-L98

It is indeed confusing, and I did miss it. Maybe there is a better way to implement this, e.g. by receiving a context or so? Not sure if we should fix this in our PR

Reply:
Where possible we should have the open and corresponding close in the same scope for these things. If it's not a major refactor we should consider it. I wouldn't mind pairing on it if you want.

Reply:
I think this can be implemented by injecting a `Context` into that rows object that would be closed in the `Close()`, and then starting a go routine waiting for that context to be done to call Rollback. I will give it a go now.

Reply:
@blairboy362 the idea is this: ef278152bfdb544bae4d46a53e84b576e040d18d

What do you think?

The only doubt I have is that now `rows.Close()` won't return an error. I wonder if it is worth it to have a way to get the error back or not.

Reply:
Other implementation can be: 9bd7ee45dc76a53e33ae63b1252afb001ac06b06

That simple uses a callback to be called from `billableRows.Close()`. That way we define the `tx.Rollback()` and `rows.Close()` next to the creation of the transaction.